test/test_genome.py

```diff
@@ -346,89 +347,118 @@ def test_is_gene_in_output_region(rng_seed):
+    only_silent_mutations_1 = genome.mutate(mutation_rate=1, rng=rng)
+    assert not only_silent_mutations_1
+
+
+# Overwrite global genome_params() from conftest,
+# since tests below strongly depend on these variable values
+@fixture
+def genome_params():
+    return {
+        "n_inputs": 2,
+        "n_outputs": 1,
+        "n_columns": 3,
+        "n_rows": 3,
+        "levels_back": 2,
+        "primitives": (cgp.Add, cgp.Sub, cgp.ConstantFloat),
+    }
+
+
+def test_permissible_values_hidden(genome_params, rng_seed):
+    genome = cgp.Genome(**genome_params)
+
+    # test function gene
+    gene_idx = 6  # function gene of first hidden gene
+    gene = 0  #
+    region_idx = 1000  # can be any number, since not touched for function gene
+    permissible_function_gene_values = genome._determine_permissible_values_hidden(
+        gene_idx, gene, region_idx
+    )
+
+    # test input gene
+    gene_idx = 16  # first input gene in second hidden layer
+    gene = 0
+    region_idx = 5
+    permissible_hidden_input_gene_values = genome._determine_permissible_values_hidden(
+        gene_idx, gene, region_idx
+    )
+
+    assert permissible_function_gene_values == [
+        1,
         2,
-        ID_NON_CODING_GENE,
-    ]
+    ]  # function idx 1,2 (cgp.Sub, cgp.ConstantFloat)
+    assert permissible_hidden_input_gene_values == [
+        1,
+        2,
+        3,
+        4,
+    ]  # gene indices of input, 1st hidden layer
 
-    assert genome._mutate_output_region(list(dna), 9, rng) is True
-    assert genome._mutate_output_region(list(dna), 10, rng) is False
-    assert genome._mutate_output_region(list(dna), 11, rng) is True
 
+def test_permissible_values_output(genome_params, rng_seed):
+    genome = cgp.Genome(**genome_params)
 
-@pytest.mark.parametrize("mutation_rate", [0.02, 0.05, 0.2])
-def test_correct_number_of_mutations(mutation_rate, rng_seed):
+    gene_idx_0 = 33
+    gene_idx_1 = 34
+    gene_idx_2 = 35
 
-    n_inputs = 2
-    n_outputs = 1
-    n_columns = 10
-    n_rows = 2
-    levels_back = 5
-    primitives = (cgp.Add, cgp.Sub, cgp.Mul, cgp.Div, cgp.ConstantFloat)
+    gene = 5  # possible input genes: 0-10
 
-    rng = np.random.RandomState(rng_seed)
-    genome = cgp.Genome(n_inputs, n_outputs, n_columns, n_rows, levels_back, primitives)
-    genome.randomize(rng)
+    permissible_values_0 = genome._determine_permissible_values_output(gene_idx_0, gene)
+    permissible_values_1 = genome._determine_permissible_values_output(gene_idx_1, gene)
+    permissible_values_2 = genome._determine_permissible_values_output(gene_idx_2, gene)
 
-    n_mutations = 0
-    genome_new = genome.clone()
-    genome_new.mutate(mutation_rate, rng)
-    for (gene_0, gene_1) in zip(genome.dna, genome_new.dna):
-        if gene_0 != gene_1:
-            n_mutations += 1
-
-    n_mutations_expected = int(mutation_rate * len(genome.dna))
-    assert n_mutations == n_mutations_expected
+    assert permissible_values_0 == []
```

Comment:
as above, i'd put the asserts right under the corresponding function calls, this way tests fail early